File: Week2_HashTables/3sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> out = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) {
                break;
            }
            if (i == 0 || nums[i] != nums[i - 1]) {
                threeSumAux(nums, i, out);
            }
        }
        return out;
    }

    public void threeSumAux(int[] nums, int i, List<List<Integer>> out) {
        int left = i + 1;
        int right = nums.length - 1;
        while (left < right) {
            int target = nums[i] + nums[left] + nums[right];
            if (target == 0) {
                out.add(Arrays.asList(nums[i], nums[left], nums[right]));
                left++;
                right--;
                while (left < right && nums[left] == nums[left - 1]) {
                    left++;
                }
            } else if (target > 0) {
                right--;
            } else {
                left++;
            }
        }
    }
}

File: Week2_HashTables/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        output = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0:
                break;
            if i == 0 or nums[i] != nums[i - 1]:
                right = len(nums) - 1
                left = i + 1
                while right > left:
                    target = nums[i] + nums[right] + nums[left]
                    if target < 0:
                        left += 1  
                    elif target > 0:
                        right -= 1
                    else:
                        output.append([nums[i], nums[left], nums[right]])
                        left += 1
                        right -= 1
                        while right > left and nums[right] == nums[right + 1]:
                            right -= 1  
        return output